### DISCUSSION TITLE
podman-compose in Windows not recognized in CLI

### DISCUSSION THREAD

**Discussion by lgonzalezsa** [podman-desktop/podman-desktop#2056] [Q&A] [answered]
### Bug description

After following the installation steps for podman and podman-compose, podman desktop indicates is installed and DOCKER_HOST is reachable but in CLI I was not able to invoke `podman-compose`

What am I missing?

![image](https://user-images.githubusercontent.com/704686/231642119-df4fea67-68b5-427a-bd27-450fdf42f900.png)

### Operating system

Windows 10

### Version

0.13.0

**Comment by benoitf** [comment]
Hello, the compose that is being fetched is docker-compose

so running 'compose' from the path displayed by the 'compose' status bar item should work


if you want to use python podman-compose you'll need to use https://github.com/containers/podman-compose#installation

**Reply by lgonzalezsa** [reply]
Thanks for the answer. How can I discover the path where `compose` is located, was trying hover over Compose label and checking in preferences but no luck. See video attached.

https://user-images.githubusercontent.com/704686/231780593-c63fd649-6278-4e7b-9289-ac329f5ec613.mp4

**Reply by cdrage** [reply]
Hi @lgonzalezsa !

I know why this is confusing now.

When we install "compose" through Podman Desktop, it's actually using the `docker-compose` CLI with a small wrapper around it.

We do not use https://github.com/containers/podman-compose (the python implementation), hence the confusion.

When we install the docker-compose binary, it's installed on windows to your `(USERDIR)/AppData/Roaming/containers` folder, where you should be able to find the podman-desktop folder with docker-compose in it.

You are right, it should be installed system-wide and accessible through the CLI, but it's not. We have implemented this with `kind`, but not yet with compose. We have opened an issue up here: https://github.com/containers/podman-desktop/issues/1974

**Reply by lgonzalezsa** [reply]
Thank you @cdrage, was checking the VM am using for testing and notice there is no folder inside of `(USERDIR)/AppData/Roaming/containers`, just `containers.conf`
Is that expected?
![image](https://user-images.githubusercontent.com/704686/233498300-ff907970-e14b-47a5-925e-10d97db7df66.png)

**Reply by davdr** [reply]
It's installed in `%USERPROFILE%\.local\share\containers\podman-desktop\extensions-storage\compose\bin\docker-compose.exe`

**Reply by davdr** [reply]
@benoitf that said, [extensions/compose/src/compose-wrapper-generator.ts](https://github.com/containers/podman-desktop/blob/e884bc896444a9dea7ad1c4ac257462155dc2439/extensions/compose/src/compose-wrapper-generator.ts) leads me to think a `podman-compose.bat` wrapper around `%USERPROFILE%\.local\share\containers\podman-desktop\extensions-storage\compose\bin\docker-compose.exe` should be generated, but it's not?

**Reply by benoitf** [reply]
@davdr you should have a compose.bat

**Reply by davdr** [reply]
@benoitf I looked for it but I can't find it I'm afraid. In which directory should it have been installed?

**Reply by davdr** [reply]
For completeness:

```
PS C:\Users\davdr> dir -Path $env:USERPROFILE -Filter "*podman*.exe" -Recurse -ErrorAction SilentlyContinue -Force | %{$_.FullName}
C:\Users\davdr\AppData\Local\Programs\podman-desktop\Podman Desktop.exe
C:\Users\davdr\AppData\Local\Programs\podman-desktop\Uninstall Podman Desktop.exe
C:\Users\davdr\AppData\Local\Programs\podman-desktop\resources\extensions\podman\assets\podman-4.4.4-setup.exe
```

```
PS C:\Users\davdr> dir -Path $env:USERPROFILE -Filter "*compose*.exe" -Recurse -ErrorAction SilentlyContinue -Force | %{$_.FullName}
C:\Users\davdr\.local\share\containers\podman-desktop\extensions-storage\compose\bin\docker-compose.exe
```

```
PS C:\Users\davdr> dir -Path $env:USERPROFILE -Filter "*compose*.bat" -Recurse -ErrorAction SilentlyContinue -Force | %{$_.FullName}
(no results)
```

I have Podman Desktop v0.14.1.

**Reply by benoitf** [reply]
@davdr could you remove the C:\Users\davdr\.local\share\containers\podman-desktop\extensions-storage\compose\bin\docker-compose.exe binary, then click again on the compose icon in the status bar
and try to install again
Also, open the Console view through the View/Toggle Developer Tools Menu
(Left alt key should display the menu bar if not shown automatically)

**Reply by davdr** [reply]
@benoitf removed binary, reinstalled, nothing additional appeared in the logs (I displayed "All levels", but do I need to increase the log level used somehow?). I did get the popup window saying the executable was downloaded.

Before I also already tried reinstalling completely (Podman Desktop and Podman removed, directories removed, reinstalled Podman Desktop and let it download and install Podman), no change.

Maybe we should turn this into an issue instead of discussion?

```
index.ts:1259 main ↪️ Kubeconfig path C:\Users\dadr\.kube\config provided does not exist. Skipping.
(anonymous) @ index.ts:1259
index.ts:1259 main ↪️ PluginSystem: received dom-ready event from the UI
index.ts:1259 main ↪️ Delayed startup, flushing 6 events
index.ts:1259 main ↪️ System ready. Loading extensions...
index.ts:1259 main ↪️ Activating extension (compose)
index.ts:1259 main ↪️ Activation extension (compose) ended
index.ts:1259 main ↪️ Activating extension (kube-context)
index.ts:1259 main ↪️ [kube-context] starting extension kube-context
index.ts:1259 main ↪️ Activation extension (kube-context) ended
index.ts:1259 main ↪️ Activating extension (podman)
index.ts:1259 main ↪️ Activating extension (registries)
index.ts:1259 main ↪️ Activation extension (registries) ended
index.ts:1259 main ↪️ Activating extension (docker)
index.ts:1259 main ↪️ Activation extension (docker) ended
index.ts:1259 main ↪️ Activating extension (kind)
index.ts:1259 main ↪️ Activating extension (lima)
index.ts:1259 main ↪️ [lima] Could not find podman socket at C:\Users\dadr\.lima\podman-lima\sock\podman.sock nor C:\Users\dadr\.lima\podman-lima\sock\podman.sock
(anonymous) @ index.ts:1259
index.ts:1259 main ↪️ Activation extension (lima) ended
index.ts:1259 main ↪️ (node:9916) UnhandledPromiseRejectionWarning: Error: EINVAL: invalid argument, stat '//./pipe/'
(Use `Podman Desktop --trace-warnings ...` to show where the warning was created)
(anonymous) @ index.ts:1259
index.ts:1259 main ↪️ (node:9916) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(anonymous) @ index.ts:1259
index.ts:1259 main ↪️ [kind] Error: '"kind-windows-amd64"' is not recognized as an internal or external command,
operable program or batch file.

    at ChildProcess.<anonymous> (C:\Users\dadr\AppData\Local\Programs\podman-desktop\resources\app.asar\extensions\kind\builtin\kind.cdix\dist\extension.js:7:11544)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
(anonymous) @ index.ts:1259
index.ts:1259 main ↪️ Activation extension (kind) ended
index.ts:1259 main ↪️ Activation extension (podman) ended
index.ts:1259 main ↪️ PluginSystem: initialization done.
index.ts:1259 main ↪️ Autostarting container engine
index.ts:1259 main ↪️ [podman] Podman extension: Autostarting machine podman-machine-default
# docker-compose downloaded here, nothing logged anymore
```

**Comment by lgonzalezsa** [comment]
Following up on this topic, then now I am using podman desktop 1.0.0 I see the docker-compose.exe binary installed.
What should I expect my Windows users to do? I need to instruct them to find the location of the binary for them to use the compose feature? Maybe I could instruct to put the path in the path variable.
I did not find any `podman-compose.bat` file nor a PATH displayed in the Compose status bar.

![image](https://github.com/containers/podman-desktop/assets/704686/d72726c1-e068-42b9-8026-6710114559ee)


https://github.com/containers/podman-desktop/issues/2055#issuecomment-1506447655

**Reply by Orca75** [reply]
In my installation its in c:\Users\<username>\.local\share\containers\podman-desktop\extensions-storage\podman-desktop.compose\bin\docker-compose.exe. 
1) it's named docker-compose, not podman-compose.
2) why it's not installed in %appdata%\Local => c:\Users\<username>\AppData\Local?

**Reply by benoitf** [reply]
Since 1.1.0 version I think it's asking to install it system wide (so it's available in your PATH)

**Reply by lgonzalezsa** [reply]
First I updated my existing installation to 1.1.0, docker-compose not in PATH.
Then, uninstalled podman-desktop and reinstall to see if the behavior changes but no luck, need to manually add to PATH.

![image](https://github.com/containers/podman-desktop/assets/704686/62c084cc-3575-49d4-9c99-e3b9625357c6)

**Reply by davdr** [reply]
Judging from [the code](https://github.com/containers/podman-desktop/pull/2718/commits/5172e7019c709bf3869a32181dbe2fc7aee548ac#diff-4a5cad339d36b12d0096e65118b2cf1bf38a4716bd1516fd6edc704453d385a6R146) the "system-wide" install should put it in 'AppData\Local\Microsoft\WindowsApps'. It doesn't add anything to the PATH, is that directory maybe on the PATH already? I don't have a Windows machine at hand at this moment, can't check on mine.

**Comment by lgonzalezsa** [comment]
This is what I see in my PATH variable, maybe an extra semicolon that should not be a problem, correct?
```C:\Users\gonluisr\AppData\Local\Microsoft\WindowsApps;;```
![image](https://github.com/containers/podman-desktop/assets/704686/257a6b0e-076b-4c96-ba34-4c3161b7cea8)

**Reply by davdr** [reply]
No, the extra semicolon is harmless I think. And is the executable in that directory? If not maybe you didn't choose "system-wide"?

**Comment by lgonzalezsa** [comment]
well, good point. The installer did not ask me anything. I did it using: 
```winget install -e --id RedHat.Podman-Desktop```
And the directory does not have something related to `docker-compose`
![image](https://github.com/containers/podman-desktop/assets/704686/6b451df5-638b-4454-89d2-7a14259cd001)

**Reply by jeffmaury** [reply, marked as answer]
Uninstalling Podman Desktop does not uninstall the configuration data so I suspect you already add compose installed but not added to path (that was added recently) so it is not being in the PATH.

Can you try to:

1. Stop Podman Desktop
2. Remove $HOME/.local/share/containers/podman-desktop/extensions-storage/podman-desktop.compose/bin
3. Restart Podman Desktop

It should ask to setup Compose and you should be good now

**Comment by lgonzalezsa** [comment]
Awesome! that made the trick! thank you!
![image](https://github.com/containers/podman-desktop/assets/704686/280070b4-a02b-46a8-b117-7a24ce8c62c6)
